File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque

def solution(board):
    answer = 0
    n = len(board)
    m = len(board[0])
    visited = [[False] * m for _ in range(n)]
    
    def find_start():
        for i in range(n):
            for j in range(m):
                if board[i][j] == "R":
                    return i, j
    
    def bfs(x,y):
        q = deque()
        q.append((x,y,0))
        visited[x][y] = True
        while q:
            cx,cy,cc = q.popleft()
            
            if board[cx][cy] == "G":
                return cc
            
            #상
            i = cx
            while i-1>=0 and board[i-1][cy] != "D":
                i -= 1
            if not visited[i][cy]:
                q.append((i, cy, cc+1))
                visited[i][cy] = True
            
            #하
            i = cx
            while i+1<n and board[i+1][cy] != "D":
                i += 1
            if not visited[i][cy]:
                q.append((i, cy, cc+1))
                visited[i][cy] = True
            
            #좌
            j = cy
            while j-1>=0 and board[cx][j-1] != "D":
                j -= 1
            if not visited[cx][j]:
                q.append((cx, j, cc+1))
                visited[cx][j] = True
            
            #우
            j = cy
            while j+1<m and board[cx][j+1] != "D":
                j += 1
            if not visited[cx][j]:
                q.append((cx, j, cc+1))
                visited[cx][j] = True
        return -1
    
    stx, sty = find_start()
    
    return bfs(stx,sty)

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cpp
#include <string>
#include <vector>
#include <queue>
#include <functional>
#include <tuple>

using namespace std;

int solution(vector<string> board) {
    int answer = 0;
    int n = board.size();
    int m = board[0].size();
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    
    function<pair<int,int>()> find_start = [&]() {
        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++){
                if (board[i][j] == 'R') {
                    return pair<int,int>(i,j);
                }
            }
        }
        return pair<int,int>(-1, -1);
    };
    
    function<int(int,int)> bfs = [&](int x, int y) {
        queue<tuple<int,int,int>> q;
        q.push({x,y,0});
        visited[x][y] = true;
        
        while (!q.empty()) {
            auto [curx,cury,curc] = q.front();
            q.pop();
            
            if (board[curx][cury] == 'G') {
                return curc;
            }
            
            // 상
            int i = curx;
            while (i-1>=0 && board[i-1][cury] != 'D') i--;
            if (!visited[i][cury]) {
                q.push({i,cury,curc+1});
                visited[i][cury] = true;
            }
            
            // 하
            i = curx;
            while (i+1<n && board[i+1][cury] != 'D') i++;
            if (!visited[i][cury]) {
                q.push({i,cury,curc+1});
                visited[i][cury] = true;
            }
            
            // 좌
            int j = cury;
            while (j-1>=0 && board[curx][j-1] != 'D') j--;
            if (!visited[curx][j]) {
                q.push({curx,j,curc+1});
                visited[curx][j] = true;
            }
            
            // 우
            j = cury;
            while (j+1<m && board[curx][j+1] != 'D') j++;
            if (!visited[curx][j]) {
                q.push({curx,j,curc+1});
                visited[curx][j] = true;
            }
        }
        return -1;
    };
    
    auto [startx,starty] = find_start();
    answer = bfs(startx,starty);
    return answer;
}